staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/oidc.go

```diff
@@ -109,6 +125,8 @@ func (c *clientCache) setClient(clusterAddress, issuer, clientID string, client
 }
 
 func newOIDCAuthProvider(clusterAddress string, cfg map[string]string, persister restclient.AuthProviderConfigPersister) (restclient.AuthProvider, error) {
+	newOIDCAuthProviderMutex.Lock()
+	defer newOIDCAuthProviderMutex.Unlock()
```

Comment:
Actually you do need both as locks in Golang are not reentrant (the same lock can not be acquired by the same Go routine twice). If I acquired a `cache.mu` lock here and then call `cache.deleteClient`, it will block forever (both will deadlock). So unless we have a reentrant lock implementation somewhere, we just need another lock.

And I do agree with you, if you have more than one lock, you need to be careful with locking order. A rule of thumb - acquire and release lock always in the same order.

So I'm open for suggestions how to improve this piece.